File: frontend/src/types/resman-preview.ts
/**
 * ResMan import preview — types mirror `app/schemas/resman_preview.py`.
 *
 * Shape:
 *   ResmanPreviewResponse
 *     ├ columns:       PreviewColumn[]    (template column + detected role)
 *     ├ rows:          PreviewRow[]       (each row's cells, in column order)
 *     ├ contributions: PreviewContribution[] ("what each report adds")
 *     └ notes:         string[]           (human-readable status / caveats)
 *
 * Cell statuses:
 *   "ok"          — value resolved (from invoice or reference report)
 *   "unresolved"  — column has a known role but no value can be filled yet
 *   "unmapped"    — column doesn't match any known role (rendered with a
 *                   subtle slash pattern so it's visually distinct from
 *                   "unresolved" — that's a missing value, this is a
 *                   missing rule)
 */

export type CellStatus = "ok" | "unresolved" | "unmapped";

export type CellSource =
  | "vendor_report"
  | "property_report"
  | "unit_report"
  | "invoice"
  | "synthetic";

export type ReferenceSource =
  | "vendor_report"
  | "property_report"
  | "unit_report"
  | "import_template";

export type ColumnRole =
  | "vendor_name"
  | "vendor_id"
  | "property_name"
  | "property_code"
  | "unit"
  | "location"
  | "invoice_number"
  | "invoice_date"
  | "due_date"
  | "amount_total"
  | "amount_subtotal"
  | "amount_tax"
  | "currency"
  | "description"
  | "account_number"
  | "gl_code"
  | "unmapped";

export type RowOrigin = "invoice" | "synthetic";

export interface PreviewCell {
  value: string | null;
  status: CellStatus;
  source: CellSource | null;
  note: string | null;
}

export interface PreviewRow {
  /** Row label shown in the leftmost (frozen) column. */
  label: string;
  origin: RowOrigin;
  /** One cell per column, in template column order. */
  cells: PreviewCell[];
}

export interface PreviewColumn {
  name: string;
  role: ColumnRole;
  /** Which uploaded report (if any) backs this column. */
  populated_from: ReferenceSource | null;
  /**
   * True when this column's role was set by a user-saved Import Builder
   * override (rather than the auto-classifier). The Import Builder
   * decorates the column header so the user can see which roles they've
   * manually pinned.
   */
  role_overridden: boolean;
}

export interface PreviewContribution {
  source: ReferenceSource;
  label: string;
  /** True when this source's file has been uploaded. */
  available: boolean;
  /** Total parsed rows in the source (NOT the sample size). */
  row_count: number | null;
  /** Template column names this source helps fill. */
  columns_powered: string[];
  note: string | null;
}

export interface ResmanPreviewResponse {
  has_template: boolean;
  template_filename: string | null;
  columns: PreviewColumn[];
  rows: PreviewRow[];
  contributions: PreviewContribution[];
  notes: string[];
}

File: backend/app/schemas/resman_preview.py
"""
Pydantic shapes for the ResMan import preview API.

Mirrors the dataclasses in `app.domain.resman_preview` but adds the
typing constraints the FastAPI layer needs (Literals for enums, default
factories for lists). The API layer converts domain dataclasses into
these schemas at the response edge.
"""

from typing import Literal

from pydantic import BaseModel, Field

CellStatusLiteral = Literal["ok", "unresolved", "unmapped"]
CellSourceLiteral = Literal[
    "vendor_report",
    "property_report",
    "unit_report",
    "invoice",
    "synthetic",
]
ReferenceSourceLiteral = Literal[
    "vendor_report",
    "property_report",
    "unit_report",
    "import_template",
]
ColumnRoleLiteral = Literal[
    "vendor_name",
    "vendor_id",
    "property_name",
    "property_code",
    "unit",
    "location",
    "invoice_number",
    "invoice_date",
    "due_date",
    "amount_total",
    "amount_subtotal",
    "amount_tax",
    "currency",
    "description",
    "account_number",
    "gl_code",
    "unmapped",
]
RowOriginLiteral = Literal["invoice", "synthetic"]


class PreviewCellOut(BaseModel):
    """One cell in the preview grid."""

    value: str | None = None
    status: CellStatusLiteral
    source: CellSourceLiteral | None = None
    note: str | None = None


class PreviewRowOut(BaseModel):
    """One row — either a real approved invoice or a synthetic sample."""

    label: str
    origin: RowOriginLiteral
    cells: list[PreviewCellOut] = Field(default_factory=list)


class PreviewColumnOut(BaseModel):
    """Template column with its detected role and the report that backs it."""

    name: str
    role: ColumnRoleLiteral
    populated_from: ReferenceSourceLiteral | None = None
    # True when an Import Builder config pinned this column's role
    # (rather than the auto-classifier deriving it). The UI uses this
    # to decorate the column header so the user can see which roles
    # they've manually overridden.
    role_overridden: bool = False


class PreviewContributionOut(BaseModel):
    """
    What one uploaded reference source contributes to the preview.

    `columns_powered` is the list of template column names this source
    helps fill. For the import template itself it's the full column
    list (the template *defines* the shape). For the three reports it's
    the subset of template columns whose detected role maps back to that
    report.
    """

    source: ReferenceSourceLiteral
    label: str
    available: bool
    row_count: int | None = None
    columns_powered: list[str] = Field(default_factory=list)
    note: str | None = None


class ResmanPreviewResponse(BaseModel):
    """
    Top-level response from GET /reference-data/preview.

    Always returns a structurally complete response — even when the
    template hasn't been uploaded yet, `columns` / `rows` are empty
    lists rather than missing keys, so the UI can render a single
    consistent skeleton + empty-state copy.
    """

    has_template: bool
    template_filename: str | None = None
    columns: list[PreviewColumnOut] = Field(default_factory=list)
    rows: list[PreviewRowOut] = Field(default_factory=list)
    contributions: list[PreviewContributionOut] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)
